### PR TITLE
Fix handling of wildcards and add extensive tests

### LINKED ISSUE #402: Allow module wildcards
In large projects you often want to enforce explicit silos of modules. These layers reduce the chance of circular dependencies and keep things a bit more sane. The module restriction feature could work perfectly for this if it had wildcards.

```
- modules:
  - {name: [Foo.Bar.*], within: [Foo.Bar.*, Foo]}
  - {name: [Foo.Qux.*], within: [Foo.Qux.*, Foo]}
```

This is likely a weaker form of the intention behind https://github.com/ndmitchell/lasagna

### PR DESCRIPTION
**Pull request by tfausak**
This builds off my previous work in #1149 and #1221, both of which aimed to fix #402. Hopefully this PR can be the last in the series 😆 

Recently I was trying to use module wildcards with HLint, both for matching imported modules and for matching which modules rules should apply to. While the former worked, I was distressed to see that the latter did not. This confused me since I thought I had implemented this feature myself. Turns out I never tested it! 

This PR fixes module wildcard handling and adds tests for all the cases I can think of right now.

### CODE REVIEW

**Comment by tfausak** [comment]
@ndmitchell can you review this? The diff may suggest that this is a big change, but really only a couple lines of code changed. Everything else is tests to make sure that the wildcard matches work properly in a variety of situations.

File: tests/wildcards.test [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+RUN tests/wildcard-xr.hs --hint=data/wildcards.yaml
+FILE tests/wildcard-xr.hs
+module XR where x = read ""
+OUTPUT
+No hints
+
+----
+RUN tests/wildcard-x-r.hs --hint=data/wildcards.yaml
+FILE tests/wildcard-x-r.hs
+module X.R where x = read ""
```
Comment (zliu41) [inline tests/wildcards.test:675]: So `*R` doesn't match `X.R` - is this intentional? It seems counter intuitive.

File: data/wildcards.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+- modules:
+  - { name: A, as: A }
+  - { name: '*B', as: B }
+  - { name: '**.C', as: C }
+  - { name: '**.*D', as: D }
+  - { name: E, within: E }
+  - { name: F, within: '*F' }
+  - { name: G, within: '**.G' }
+  - { name: H, within: '**.*H' }
```
Comment (zliu41) [inline data/wildcards.yaml:9]: Perhaps add a rule where both `name` and `within` contain wildcards?

File: tests/wildcards.test (hunk 1 above)
Comment (tfausak) [reply tests/wildcards.test:675]: Yes, that's intentional. The module wildcard matching follows the same rules as file globbing. `*R` would match `R` and `XR` but not `X.R`. Along the same lines, `**.R` would match `R` and `X.R` but not `XR`.

Comment (tfausak) [reply tests/wildcards.test:736]: Ah yeah, I can see why that's confusing. The module matching is like file globbing, but importantly `.` characters in module names are treated like `/` characters in file paths. 

```
-- file glob
**/*Spec.hs
  Spec.hs -- match
  FooSpec.hs -- match
  Foo/BarSpec.hs -- match

-- module glob
**.*Spec
  Spec -- match
  FooSpec -- match
  Foo.BarSpec -- match

**Comment by ndmitchell** [comment]
Thanks for the awesome tests, and sorry it took so long to review!

**Comment by tfausak** [comment]
No problem at all. Thanks!

**Pull request merged**